File: packages/types/src/config/config-index.ts
/**
 * Configuration Index Types
 *
 * Defines types for configuration index files that enable efficient discovery
 * and loading of large numbers of configuration files.
 *
 * Design Principles:
 * - Index files contain ONLY path mappings (minimal and simple)
 * - Metadata is defined in individual config files (single source of truth)
 * - Supports glob patterns for flexible path matching
 * - Index is just a "pointer" to actual config files
 *
 * Example index.json:
 * ```json
 * {
 *   "version": "1.0",
 *   "type": "llm_profiles",
 *   "paths": [
 *     "./openai/*.toml",
 *     "./anthropic/*.toml",
 *     "./gemini-native/*.toml"
 *   ]
 * }
 * ```
 */

/**
 * Base configuration index file structure (simplified)
 * Contains only path mappings - metadata comes from individual files
 */
export interface ConfigIndexFile {
  /** Index file format version */
  version: string;
  /** Index type identifier */
  type: IndexType;
  /**
   * Path patterns to scan for config files.
   * Supports glob patterns:
   * - `./*.toml` - all TOML files in current directory
   * - `./subdir/*.toml` - all TOML files in subdirectory
   * - `./**` + `/*.toml` - all TOML files recursively
   * - `./openai/gpt-*.toml` - specific pattern matching
   */
  paths: string[];
}

/**
 * Resolved index entry with metadata extracted from config file
 * This is the runtime representation after loading and parsing
 */
export interface ResolvedIndexEntry {
  /** Unique identifier from the config file */
  id: string;
  /** Human-readable name from the config file */
  name?: string;
  /** Description from the config file */
  description?: string;
  /** Tags from the config file */
  tags?: string[];
  /** Category from the config file */
  category?: string;
  /** Absolute path to the config file */
  filePath: string;
  /** File format (toml or json) */
  format: "toml" | "json";
  /** Additional metadata from the config file */
  metadata?: Record<string, unknown>;
}

/**
 * LLM Profile specific resolved entry
 */
export interface ResolvedLLMProfileEntry extends ResolvedIndexEntry {
  /** Provider type */
  provider?: string;
  /** Model identifier */
  model?: string;
}

/**
 * Workflow specific resolved entry
 */
export interface ResolvedWorkflowEntry extends ResolvedIndexEntry {
  /** Workflow type */
  type?: string;
  /** Workflow version */
  version?: string;
  /** Author */
  author?: string;
}

/**
 * Node Template specific resolved entry
 */
export interface ResolvedNodeTemplateEntry extends ResolvedIndexEntry {
  /** Node type */
  type?: string;
}

/**
 * Script specific resolved entry
 */
export interface ResolvedScriptEntry extends ResolvedIndexEntry {
  /** Script category */
  category?: string;
  /** Executor type */
  executor?: string;
}

/**
 * Resolved index with all entries loaded
 */
export interface ResolvedIndex<T extends ResolvedIndexEntry = ResolvedIndexEntry> {
  /** Index type */
  type: IndexType;
  /** All resolved entries */
  entries: T[];
  /** Index metadata */
  metadata?: {
    /** When the index was resolved */
    resolvedAt: string;
    /** Total count of entries */
    totalCount: number;
    /** Files that failed to load */
    failures: Array<{ path: string; error: string }>;
  };
}

/**
 * Supported index types
 */
export type IndexType =
  | "llm_profiles"
  | "workflows"
  | "node_templates"
  | "trigger_templates"
  | "hook_templates"
  | "scripts"
  | "prompt_templates"
  | "agent_loops"
  // Preset-based index types
  | "mcp_presets"
  | "skill_presets"
  | "infrastructure_presets";

/**
 * Index file names for each type
 */
export const INDEX_FILE_NAMES: Record<IndexType, string> = {
  llm_profiles: "index.json",
  workflows: "index.json",
  node_templates: "index.json",
  trigger_templates: "index.json",
  hook_templates: "index.json",
  scripts: "index.json",
  prompt_templates: "index.json",
  agent_loops: "index.json",
  mcp_presets: "index.json",
  skill_presets: "index.json",
  infrastructure_presets: "index.json",
} as const;

/**
 * Default config directories for each type
 */
export const DEFAULT_CONFIG_DIRS: Record<IndexType, string> = {
  llm_profiles: "configs/llm-profiles",
  workflows: "configs/workflows",
  node_templates: "configs/node-templates",
  trigger_templates: "configs/trigger-templates",
  hook_templates: "configs/hook-templates",
  scripts: "configs/scripts",
  prompt_templates: "configs/prompt-templates",
  agent_loops: "configs/agent-loops",
  mcp_presets: "configs/mcp",
  skill_presets: "configs/skills",
  infrastructure_presets: "configs/infrastructure",
} as const;

// ---------------------------------------------------------------------------
// Preset-related Types
// ---------------------------------------------------------------------------

/**
 * Skill collection definition file (multi-file preset)
 * Contains paths to actual skill definition files.
 */
export interface SkillCollectionFile {
  /** Collection identifier (matches filename without extension) */
  id: string;
  /** Human-readable name */
  name?: string;
  /** Description */
  description?: string;
  /** Tags for filtering */
  tags?: string[];
  /**
   * Path patterns to scan for skill definition files.
   * Supports same glob patterns as ConfigIndexFile.paths.
   */
  paths: string[];
}

/**
 * Infrastructure preset definition file (multi-file preset)
 * Maps each config domain to its file path.
 */
export interface InfrastructurePresetFile {
  /** Preset identifier (matches filename without extension) */
  id: string;
  /** Human-readable name */
  name?: string;
  /** Description */
  description?: string;
  /** Tags for filtering */
  tags?: string[];
  /** File path mapping for each config domain */
  files: {
    metrics?: string;
    timeout?: string;
    storage?: string;
    output?: string;
    fileCheckpoint?: string;
    presets?: string;
    /** Tool-specific configuration file (readFile, writeFile, etc.) */
    tools?: string;
  };
}

File: crates/wf-types/src/config/config_index.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum IndexType {
    LlmProfiles,
    Workflows,
    NodeTemplates,
    TriggerTemplates,
    Scripts,
    PromptTemplates,
    AgentLoops,
    McpPresets,
    SkillPresets,
    InfrastructurePresets,
}

impl IndexType {
    pub fn as_str(&self) -> &'static str {
        match self {
            IndexType::LlmProfiles => "llm_profiles",
            IndexType::Workflows => "workflows",
            IndexType::NodeTemplates => "node_templates",
            IndexType::TriggerTemplates => "trigger_templates",
            IndexType::Scripts => "scripts",
            IndexType::PromptTemplates => "prompt_templates",
            IndexType::AgentLoops => "agent_loops",
            IndexType::McpPresets => "mcp_presets",
            IndexType::SkillPresets => "skill_presets",
            IndexType::InfrastructurePresets => "infrastructure_presets",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConfigFileFormat {
    Toml,
    Json,
}

/// Index file schema matching the `index.json` format: `version` +
/// `type` + `paths`.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ConfigIndexFile {
    pub version: String,
    #[serde(rename = "type")]
    pub index_type: IndexType,
    pub paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResolvedIndexEntry {
    pub id: String,
    pub name: Option<String>,
    pub description: Option<String>,
    pub tags: Option<Vec<String>>,
    pub category: Option<String>,
    pub file_path: String,
    pub format: ConfigFileFormat,
    pub metadata: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResolvedLlmProfileEntry {
    #[serde(flatten)]
    pub base: ResolvedIndexEntry,
    pub provider: Option<String>,
    pub model: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResolvedWorkflowEntry {
    #[serde(flatten)]
    pub base: ResolvedIndexEntry,
    pub workflow_type: Option<String>,
    pub version: Option<String>,
    pub author: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResolvedNodeTemplateEntry {
    #[serde(flatten)]
    pub base: ResolvedIndexEntry,
    pub node_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResolvedScriptEntry {
    #[serde(flatten)]
    pub base: ResolvedIndexEntry,
    pub category: Option<String>,
    pub executor: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResolvedIndexMetadata {
    pub resolved_at: String,
    pub total_count: usize,
    pub failures: Vec<IndexLoadFailure>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct IndexLoadFailure {
    pub path: String,
    pub error: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResolvedIndex {
    pub index_type: IndexType,
    pub entries: Vec<ResolvedIndexEntry>,
    pub metadata: Option<ResolvedIndexMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SkillCollectionFile {
    pub id: String,
    pub name: Option<String>,
    pub description: Option<String>,
    pub tags: Option<Vec<String>>,
    pub paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InfrastructurePresetFile {
    pub id: String,
    pub name: Option<String>,
    pub description: Option<String>,
    pub tags: Option<Vec<String>>,
    pub files: InfrastructurePresetFiles,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct InfrastructurePresetFiles {
    pub metrics: Option<String>,
    pub timeout: Option<String>,
    pub storage: Option<String>,
    pub output: Option<String>,
    pub file_checkpoint: Option<String>,
    pub presets: Option<String>,
    pub tools: Option<String>,
    pub sandbox: Option<String>,
}
